File: src/engine/adapters/sqs/mutate.ts
import {
  DeleteMessageCommand,
  type MessageAttributeValue,
  SendMessageCommand,
  type SQSClient,
} from '@aws-sdk/client-sqs';
import type { MutationPlan, MutationResult, MutationRowOp } from '../../../shared/domain/mutations';
import type { OpCtx } from '../adapter';
import { AdapterError, assertWritable } from '../errors';
import { mapError } from './errors';

// Sentinel key (mirrors mongo/mutate.ts's `$document`, kafka/produce.ts's `$body`): a new
// message is expressed through the existing relational-shaped MutationRowOp's `values` rather
// than widening the shared mutation schema.
const BODY_FIELD = '$body';
/** Task #61: mirrors kafka/produce.ts's own `$headers` sentinel exactly — JSON-encoded
 *  `Record<string, string>`, mapped onto SendMessage's `MessageAttributes` shape below. Sent
 *  attributes round-trip back through sqs/read.ts's `pushMessage`, which already reads
 *  `message.MessageAttributes` into the row's `headers` column unchanged. */
const HEADERS_FIELD = '$headers';

function parseHeaders(raw: string | null | undefined): Record<string, string> | undefined {
  if (raw === null || raw === undefined || raw === '') return undefined;
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    throw new AdapterError('E_QUERY', 'malformed $headers JSON');
  }
  if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
    throw new AdapterError('E_QUERY', '$headers must be a JSON object of string values');
  }
  const out: Record<string, string> = {};
  for (const [k, v] of Object.entries(parsed as Record<string, unknown>)) {
    if (typeof v !== 'string') throw new AdapterError('E_QUERY', `$headers.${k} must be a string`);
    out[k] = v;
  }
  return out;
}

function toMessageAttributes(
  headers: Record<string, string> | undefined,
): Record<string, MessageAttributeValue> | undefined {
  if (!headers) return undefined;
  const out: Record<string, MessageAttributeValue> = {};
  for (const [name, value] of Object.entries(headers)) {
    out[name] = { DataType: 'String', StringValue: value };
  }
  return out;
}

/** The delete key's field name — the row's `key` column is already the MessageId (read.ts's
 *  `pushMessage`), so the renderer echoes it back unchanged, mirroring mongo/mutate.ts's `_id`. */
const ID_FIELD = 'messageId';

function renderOpText(op: MutationRowOp, queueName: string): string {
  if (op.kind === 'insert') return `SendMessage(${queueName})`;
  if (op.kind === 'delete') return `DeleteMessage(${queueName})`;
  // A delivered message can't be edited in place, only replaced by delete+resend (sqsCaps's own
  // comment) — there is no per-message update in the SQS API.
  throw new AdapterError(
    'E_UNSUPPORTED',
    'sqs has no update operation — delete and resend instead',
  );
}

/** Synchronous (Adapter rule 3): no network, no queue-URL resolution — mirrors mongo/mutate.ts. */
export function preview(plan: MutationPlan, queueName: string): string[] {
  return plan.ops.map((op) => renderOpText(op, queueName));
}

/**
 * `receiptHandles` is the adapter-local, in-memory map SqsAdapter threads through from read.ts's
 * pollQueue (keyed by MessageId) — a receipt handle is an AWS-internal token with no reason to
 * round-trip through the wire protocol (StreamPage carries no such field), and it is only ever
 * valid for the message that was actually received, not a stable identifier of the message itself.
 * Deleting a message the current session never polled (its handle isn't in the map, or the queue
 * was reconnected since) is reported as E_QUERY rather than silently doing nothing.
 */
export async function mutateQueue(
  client: SQSClient,
  queueUrl: string,
  queueName: string,
  readOnly: boolean,
  plan: MutationPlan,
  receiptHandles: Map<string, string>,
  ctx: OpCtx,
): Promise<MutationResult> {
  assertWritable(readOnly);

  ctx.setCommand(preview(plan, queueName).join(';\n'));
  let affectedRows = 0;
  try {
    for (const op of plan.ops) {
      if (op.kind === 'insert') {
        const body = op.values[BODY_FIELD];
        if (typeof body !== 'string') {
          throw new AdapterError('E_QUERY', 'a new message requires a $body');
        }
        const headers = parseHeaders(op.values[HEADERS_FIELD]);
        await client.send(
          new SendMessageCommand({
            QueueUrl: queueUrl,
            MessageBody: body,
            MessageAttributes: toMessageAttributes(headers),
          }),
          { abortSignal: ctx.signal },
        );
        affectedRows++;
      } else if (op.kind === 'delete') {
        const messageId = op.key[ID_FIELD];
        if (typeof messageId !== 'string') {
          throw new AdapterError('E_QUERY', `a delete requires the message's ${ID_FIELD}`);
        }
        const receiptHandle = receiptHandles.get(messageId);
        if (!receiptHandle) {
          throw new AdapterError(
            'E_QUERY',
            'this message was not received in the current session (its receipt handle is gone) — poll again before deleting',
          );
        }
        await client.send(
          new DeleteMessageCommand({ QueueUrl: queueUrl, ReceiptHandle: receiptHandle }),
          { abortSignal: ctx.signal },
        );
        receiptHandles.delete(messageId);
        affectedRows++;
      } else {
        throw new AdapterError(
          'E_UNSUPPORTED',
          'sqs has no update operation — delete and resend instead',
        );
      }
    }
  } catch (err) {
    if (err instanceof AdapterError) throw err;
    throw mapError(err);
  }

  return { affectedRows };
}

File: shell/internal/adapters/sqs/mutate.go
package sqs

import (
	"context"
	"encoding/json"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	"github.com/aws/aws-sdk-go-v2/service/sqs/types"

	"github.com/kirathecat/kira-studio/shell/internal/adapters"
	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

// Sentinel keys (mirrors mongo's $document, kafka's $body): a new message is expressed through
// the existing relational-shaped RowValues rather than widening the shared mutation schema.
const (
	bodyField    = "$body"
	headersField = "$headers"
	idField      = "messageId" // the row's key column is already the MessageId (read.go's pushMessage)
)

func parseHeaders(raw *string) (map[string]string, error) {
	if raw == nil || *raw == "" {
		return nil, nil
	}
	var parsed any
	if err := json.Unmarshal([]byte(*raw), &parsed); err != nil {
		return nil, adapters.New(adapters.CodeQuery, "malformed $headers JSON", err)
	}
	obj, ok := parsed.(map[string]any)
	if !ok {
		return nil, adapters.New(adapters.CodeQuery, "$headers must be a JSON object of string values", nil)
	}
	out := make(map[string]string, len(obj))
	for k, v := range obj {
		s, ok := v.(string)
		if !ok {
			return nil, adapters.New(adapters.CodeQuery, "$headers."+k+" must be a string", nil)
		}
		out[k] = s
	}
	return out, nil
}

func toMessageAttributes(headers map[string]string) map[string]types.MessageAttributeValue {
	if headers == nil {
		return nil
	}
	out := make(map[string]types.MessageAttributeValue, len(headers))
	for name, value := range headers {
		out[name] = types.MessageAttributeValue{DataType: aws.String("String"), StringValue: aws.String(value)}
	}
	return out
}

func renderOpText(op model.MutationRowOp, queueName string) (string, error) {
	switch op.Kind {
	case "insert":
		return "SendMessage(" + queueName + ")", nil
	case "delete":
		return "DeleteMessage(" + queueName + ")", nil
	default: // update — a delivered message can't be edited in place, only replaced by delete+resend.
		return "", adapters.New(adapters.CodeUnsupported, "sqs has no update operation — delete and resend instead", nil)
	}
}

// preview is mutate.ts's preview — synchronous (Adapter rule 3): no network, no queue-URL
// resolution.
func preview(plan model.MutationPlan, queueName string) ([]string, error) {
	out := make([]string, len(plan.Ops))
	for i, op := range plan.Ops {
		text, err := renderOpText(op, queueName)
		if err != nil {
			return nil, err
		}
		out[i] = text
	}
	return out, nil
}

// mutateQueue is mutate.ts's mutateQueue. handles is the adapter-local, mutex-guarded receipt-
// handle cache read.go populates — a receipt handle is only ever valid for the message that was
// actually received, not a stable identifier of the message itself; deleting a message the
// current session never polled is reported as E_QUERY rather than silently doing nothing.
func mutateQueue(ctx context.Context, client *sqs.Client, queueURL, queueName string, readOnly bool, plan model.MutationPlan, handles *receiptHandles, op *adapters.OpCtx) (model.MutationResult, error) {
	if err := adapters.AssertWritable(readOnly); err != nil {
		return model.MutationResult{}, err
	}

	statements, err := preview(plan, queueName)
	if err != nil {
		return model.MutationResult{}, err
	}
	commandText := ""
	for i, s := range statements {
		if i > 0 {
			commandText += ";\n"
		}
		commandText += s
	}
	op.SetCommand(commandText)

	affectedRows := 0
	for _, rowOp := range plan.Ops {
		switch rowOp.Kind {
		case "insert":
			body, ok := rowOp.Values.Get(bodyField)
			if !ok || body == nil {
				return model.MutationResult{}, adapters.New(adapters.CodeQuery, "a new message requires a "+bodyField, nil)
			}
			headersRaw, _ := rowOp.Values.Get(headersField)
			headers, err := parseHeaders(headersRaw)
			if err != nil {
				return model.MutationResult{}, err
			}
			_, err = client.SendMessage(ctx, &sqs.SendMessageInput{
				QueueUrl:          aws.String(queueURL),
				MessageBody:       body,
				MessageAttributes: toMessageAttributes(headers),
			})
			if err != nil {
				return model.MutationResult{}, mapError(err)
			}
			affectedRows++

		case "delete":
			messageID, ok := rowOp.Key.Get(idField)
			if !ok || messageID == nil {
				return model.MutationResult{}, adapters.New(adapters.CodeQuery, "a delete requires the message's "+idField, nil)
			}
			handle, ok := handles.get(*messageID)
			if !ok {
				return model.MutationResult{}, adapters.New(adapters.CodeQuery,
					"this message was not received in the current session (its receipt handle is gone) — poll again before deleting", nil)
			}
			_, err := client.DeleteMessage(ctx, &sqs.DeleteMessageInput{QueueUrl: aws.String(queueURL), ReceiptHandle: aws.String(handle)})
			if err != nil {
				return model.MutationResult{}, mapError(err)
			}
			handles.delete(*messageID)
			affectedRows++

		default:
			return model.MutationResult{}, adapters.New(adapters.CodeUnsupported, "sqs has no update operation — delete and resend instead", nil)
		}
	}

	return model.MutationResult{AffectedRows: affectedRows}, nil
}
